src/igraph/__init__.py

```diff
@@ -3190,6 +3190,54 @@ def DataFrame(klass, edges, directed=True, vertices=None, use_vids=False):
 
         return g
 
+    def get_vertex_dataframe(self):
+        """Export vertices with attributes to pandas.DataFrame
+
+        @return: a pandas.DataFrame representing vertices and their attributes.
+          If the vertices have names, the index of the dataframe contains them,
+          otherwise it's integers from 0 up (which are also the vertex IDs).
+          The order of the vertices is the same as in the graph, so the n-th
+          row describes the vertex with ID i.
+        """
+        import pandas as pd
+
+        df = pd.DataFrame(
+            {attr: self.vs[attr] for attr in self.vertex_attributes()},
+            index=list(range(self.vcount())),
+        )
+        if 'name' in df.columns:
+            df.set_index('name', inplace=True)
+        return df
+
+    def get_edge_dataframe(self):
+        """Export edges with attributes to pandas.DataFrame
+
+        @return: a pandas.DataFrame representing edges and their attributes.
+          If the edges have names, the index of the dataframe contains them,
+          otherwise it's integers from 0 up (which are also the edge IDs).
+          The order of the edges is the same as in the graph, so the n-th
+          row describes the edge with ID i. The first two columns of the
+          dataframe represent the IDs of source and target vertices for each
+          edge. These columns have names "source" and "target". If your
+          edges have attributes with the same names, they will be present in
+          the dataframe, but not in the first two columns.
+        """
+        import pandas as pd
+
+        df = pd.DataFrame(
+            {attr: self.es[attr] for attr in self.edge_attributes()},
+            index=list(range(self.ecount())),
+        )
+        if 'name' in df.columns:
+            df.set_index('name', inplace=True)
```

Comment:
I don't think we ever use the `name` attribute for `edges` anywhere else, right? I think it might be strange to start using it here.

Instead, I would suggest that we add the `source` and `target` columns as index.

Reply:
Hmm... yes, the current version is just me being naive about the `name` attribute.

On the other hand, I am a little worried setting up a `pd.MultiIndex` with `(source, target)` columns as index. I use `pd.MultiIndex` a fair bit in my daily work, and it's kind of a pain. Say we want to get all incoming edges into vertex `i`:

```python
df.loc[df['target'] == i]  # without multi-index
df.query('target == i')  # without multi-index
df.loc[pd.IndexSlice[:, i], :]  # with multi-index
df.loc[df.index.get_level_values('target') == i]  # with multi-index
```
To me it looks like options 1 and 2 are easier to understand and actually mention the word `target`, while option 3 uses exotic objects (`IndexSlice`), exotic constructors via square brackets, and has a hidden requirement for specification of both row and column slicers (all those colons) which can otherwise fail silently (happened before). Option 4 is easier to understand but verbose.

I should have started replying to your review here, but alas I started below:

https://github.com/igraph/python-igraph/pull/355#discussion_r558615150

so what I wrote there is that ideally we wouldn't try to baby-feed one liners to our users. Having the multiindex is such a case:

```python
df.set_index(['source', 'target'], inplace=True)  # From columns to multi-index
df.reset_index(inplace=True)  # From multi-index to column, less well known I think
```
and if they use `drop=False` they can still slice using any of the 4 options. In such cases, I'd be in favour of documenting these transformations in the docstring rather than doing them for the user.

In any case this is just a slight preference, if you are strongly opinionated I'm happy to set up the multi-index too and document the one-liner required to snap out of it.